### PR TITLE
support user specified author

### PR DESCRIPTION
**Pull request by superyyrrzz**

### CODE REVIEW

File: src/docfx/lib/git/GitRepoInfoProvider.cs [hunk 1]
```diff
@@ -36,38 +37,54 @@ public GitRepoInfoProvider(Docset docset)
             _githubUserProfileCache = GitUserProfileCache.Create(userProfileCachePath);
         }
 
-        public (GitUserInfo author, GitUserInfo[] contributors, DateTime updatedAt) GetContributorInfo(Document document)
+        public (GitUserInfo author, GitUserInfo[] contributors, DateTime updatedAt, List<Error> errors) GetContributorInfo(
+            Document document,
+            JObject metadata)
```
Comment (yufeih) [inline src/docfx/lib/git/GitRepoInfoProvider.cs]: Make input params minimal:  `(Document, string author)`

File: src/docfx/build/markdown/BuildMarkdown.cs [hunk 2]
```diff
@@ -52,6 +52,7 @@ internal static class BuildMarkdown
 
             // TODO: make build pure by not output using `context.Report/Write/Copy` here
             context.Report(file, markup.Errors);
+            context.Report(file, repoErrors);
```
Comment (yufeih) [inline src/docfx/build/markdown/BuildMarkdown.cs]: use `Concat` and call Report once

File: src/docfx/lib/git/GitRepoInfoProvider.cs [hunk 3]
```diff
@@ -36,38 +37,54 @@ public GitRepoInfoProvider(Docset docset)
             _githubUserProfileCache = GitUserProfileCache.Create(userProfileCachePath);
         }
 
-        public (GitUserInfo author, GitUserInfo[] contributors, DateTime updatedAt) GetContributorInfo(Document document)
+        public (GitUserInfo author, GitUserInfo[] contributors, DateTime updatedAt, List<Error> errors) GetContributorInfo(
```
Comment (yufeih) [inline src/docfx/lib/git/GitRepoInfoProvider.cs]: Put `errors` as the first return value

File: src/docfx/lib/git/GitRepoInfoProvider.cs [hunk 4]
```diff
@@ -36,38 +37,54 @@ public GitRepoInfoProvider(Docset docset)
...
             }
+
             var contributors = (from commit in commits
                                 where !string.IsNullOrEmpty(commit.AuthorEmail)
                                 let info = _githubUserProfileCache.GetByUserEmail(commit.AuthorEmail)
                                 where info != null
                                 group info by info.Id into g
-                                select g.First()).ToArray();
+                                select g.First()).ToList();
+            if (authorInfo != null && contributors.All(p => p.Name != authorInfo.Name))
```
[the remark this answers is not in the document]
Comment (superyyrrzz) [reply src/docfx/lib/git/GitRepoInfoProvider.cs]: identifier, e.g. `OsmondJiang`

---
In reply to: [198378743](https://github.com/dotnet/docfx/pull/2951#discussion_r198378743) [](ancestors = 198378743)

File: src/docfx/build/markdown/BuildMarkdown.cs [hunk 5]
```diff
@@ -32,7 +34,7 @@ internal static class BuildMarkdown
             var metadata = JsonUtility.Merge(Metadata.GetFromConfig(file), markup.Metadata);
             var content = markup.HasHtml ? HtmlUtility.TransformHtml(document.DocumentNode, node => node.StripTags()) : html;
             var (id, versionIndependentId) = file.Docset.Redirections.TryGetDocumentId(file, out var docId) ? docId : file.Id;
-            var (author, contributors, updatedAt, repoErrors) = repo.GetContributorInfo(file, metadata);
+            var (repoErrors, author, contributors, updatedAt) = repo.GetContributorInfo(file, GetInputAuthor(metadata));
```
Comment (yufeih) [inline src/docfx/build/markdown/BuildMarkdown.cs]: metadata.Value<string>("author")

File: src/docfx/lib/git/GitRepoInfoProvider.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/lib/git/GitRepoInfoProvider.cs]: I means is the name the only identifier?

---
In reply to: [198380401](https://github.com/dotnet/docfx/pull/2951#discussion_r198380401) [](ancestors = 198380401,198378743)

[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/lib/git/GitRepoInfoProvider.cs]: suggest to hide the identifier logic to contributor class with Equals and GetHashCode, then you can just use a HashSet

---
In reply to: [198395977](https://github.com/dotnet/docfx/pull/2951#discussion_r198395977) [](ancestors = 198395977,198380401,198378743)

[the remark this answers is not in the document]
Comment (superyyrrzz) [reply src/docfx/lib/git/GitRepoInfoProvider.cs]: I mistook it... should use Id

[the remark this answers is not in the document]
Comment (yufeih) [reply src/docfx/lib/git/GitRepoInfoProvider.cs]: can we have a unit test covering this?

---
In reply to: [198397962](https://github.com/dotnet/docfx/pull/2951#discussion_r198397962) [](ancestors = 198397962,198395977,198380401,198378743)

**Pull request merged**